samples/99_zil_interpreter/app/parser.rb

```diff
@@ -0,0 +1,403 @@
+      log indent(@expr_depth) + "+ LIST " if debug_flag
+      expr = read_list
+    elsif expr_char == '"' #  Starts with '"' then read_string
+      log indent(@expr_depth) + "+ STRING " if debug_flag
+      expr = read_string
+    elsif expr_char == '!' && peek_next == '\\' #  Starts with '!\' then read_character
+      log indent(@expr_depth) + "+ CHARACTER " if debug_flag
+      expr = read_character
+    elsif expr_char == ';' #  Starts with ';' then read_comment
+      log indent(@expr_depth) + "+ COMMENT " if debug_flag
+      expr = read_comment
+    elsif expr_char == '#' && peek_string(5) == '#DECL'
+      log indent(@expr_depth) + "+ DECL " if debug_flag
+      expr = read_decl
+    elsif expr_char == '%' && peek_string(6) == '%<COND'
+      log indent(@expr_depth) + "+ COND MACRO " if debug_flag
+      expr = read_cond_macro
+    elsif expr_char == "'"
+      log indent(@expr_depth) + "+ MACRO " if debug_flag
+      expr = read_macro
+    elsif is_fix?(expr_char) # Starts with 0123456789 then read_fix
+      log indent(@expr_depth) + "+ FIX " if debug_flag
+      expr = read_fix
+    else
+      #  else read atom
+      log indent(@expr_depth) + "+ ATOM " if debug_flag
+      expr = read_atom
+    end
+
+    @expr_depth -= 1
+    expr
+  end
+
+  def read_fix
+    # UC1: 123456
+    # UC2: *3777*
+    # UC3: #2 0111
+
+    fix = ""
+
+    # step 1: just load the string based on the three types
+    while @buf_index < @buffer.length - 1 # iterate characters in line
+      buf_char = peek
+      buffer_advance
+
+      if FIX_CHARS.include?(buf_char)
+        #log "[TMP] " + buf_char
+        fix += buf_char
+      elsif buf_char == " " && fix == '#2'
+        fix += buf_char
+      else
+        buffer_previous
+        break
+      end
+    end
+
+    # step 2: convert to number
+    if fix.start_with?('*')
+      fix = fix.gsub('*', '').to_i(base=8)
+    elsif fix.start_with?('#2')
+      fix = fix.gsub('#2 ', '').to_i(base=2)
+    else
+      fix = fix.to_i
+    end
+
+    fix
+  end
+
+  # <...>
+  def read_form
+    skip_char('<') # 1) Read '<'
+    atom = read_atom if (peek != '>') # 2) Read atom
+    skip_whitespace # 3) Skip whitespace
+
+    # 4) Read params (form, list, string, fix)
+    expr = []
+    last_buf_index = -1 # make sure we don't get into an endless loop
+    done = false
+    until done
+      if @buf_index == last_buf_index # endless loop detected
+        raise "Endless scan detected at #{@line_num}:#{@line_pos}!"
+      end
+
+      last_buf_index = @buf_index
+
+      ft = peek
+      if ft == '>' # handle special case of empty form
+        done = true
+      else
+        expr << read_expression
+        #log expr[expr.length - 1]
+      end
+
+      skip_whitespace
+    end
+
+    skip_char('>') # 5) Read '>'
+
+    if (atom.nil? || atom.length == 0) && expr.length == 0
+      Syntax::Form.new # empty Form
+    else
+      Syntax::Form.new(atom, *expr) # atom and params
+    end
+  end

```

Comment:
Reading this one more time I wonder why you even have to distinguish between `atom` and `expr` - couldn't you just use `read_list` but with `<` `>` instead of `(` `)` and outputting `Form` instead of `List`?

You could probably even extract a common helper method
```rb
def read_bracket_expression(start_bracket, end_bracket, output_class)
  skip_whitespace
  skip_char(start_bracket)

  # read contents
  expr = []
  next_char = peek
  while next_char != end_bracket
    expr << read_expression
    skip_whitespace
    next_char = peek
  end

  skip_char(end_bracket)
    
  output_class.new(*expr)
end
```

Reply:
I can experiment with this.  It seems like a potential improvement.

Reply:
This works.  It will be in next commit.